File: 백준/Silver/24954. 물약 구매/물약 구매.py
import sys
input = sys.stdin.readline

def dfs():
    global ans

    if len(arr) == n:
        #print(*arr)
        tmp = price[:]
        p = 0

        for i in arr:
            p += tmp[i]
            for ad in dic[i]:
                a, d = ad[0] - 1, ad[1]
                tmp[a] -= d
                if tmp[a] <= 0:
                    tmp[a] = 1
        ans = min(ans, p)
        return

    for i in range(n):
        if visit[i]:
            continue
        arr.append(i)
        visit[i] = True
        dfs()
        visit[i] = False
        arr.pop()


n = int(input())
price = list(map(int, input().split()))
dic = {}

for i in range(n):
    p = int(input())
    tmp = []
    for _ in range(p):
        tmp.append(list(map(int,input().split())))
    dic[i] = tmp

arr = []
ans = 1e19
visit = [False] * n
dfs()
print(ans)

File: 백준/Silver/24954. 물약 구매/물약 구매.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    static int[] arr;
    static int n;

    static boolean[] visit;
    static int[][] discount;
    static int[] sel;
    static int ans = Integer.MAX_VALUE;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        n = Integer.parseInt(br.readLine());

        arr = new int[n];
        discount = new int[n][n];
        visit = new boolean[n];

        st = new StringTokenizer(br.readLine());

        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        for (int i = 0; i < n; i++) {
            int cnt = Integer.parseInt(br.readLine());
            for (int j = 0; j < cnt; j++) {
                st = new StringTokenizer(br.readLine());
                int a = Integer.parseInt(st.nextToken());
                int b = Integer.parseInt(st.nextToken());
                discount[i][a - 1] = b;
            }
        }

        sel = new int[n];
        recursive(0);
        System.out.println(ans);
    }

    private static void recursive(int cur) {
        if(cur == n){
            int[] tmp = new int[n];
            for (int i = 0; i < n; i++) {
                tmp[i] = arr[i];
            }
            boolean[] buy = new boolean[n];
            for (int i = 0; i < n; i++) {
                buy[sel[i]] = true;
                for (int j = 0; j < n; j++) {
                    if(discount[sel[i]][j] != 0 && !buy[j]){
                        tmp[j] -= discount[sel[i]][j];
                    }
                }
            }
            int sum = 0;

            for (int i = 0; i < n; i++) {
                sum += tmp[i] <= 0 ? 1 : tmp[i];
            }

            ans = Math.min(ans, sum);
            return;
        }

        for (int i = 0; i < n; i++) {
            if(visit[i]) continue;
            visit[i] = true;
            sel[cur] = i;
            recursive(cur + 1);
            visit[i] = false;
        }
    }
}